jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java

```diff
@@ -0,0 +1,60 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.server;
+
+import java.util.function.Consumer;
+
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.handler.ContextHandler;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.websocket.api.WebSocketContainer;
+import org.eclipse.jetty.websocket.core.WebSocketComponents;
+import org.eclipse.jetty.websocket.core.server.WebSocketMappings;
+import org.eclipse.jetty.websocket.core.server.WebSocketServerComponents;
+
+public class WebSocketUpgradeHandler extends Handler.Wrapper
+{
+    public static WebSocketUpgradeHandler from(Server server, ContextHandler context)
+    {
+        WebSocketUpgradeHandler wsHandler = new WebSocketUpgradeHandler(WebSocketServerComponents.ensureWebSocketComponents(server, context));
+        ServerWebSocketContainer container = wsHandler.container;
+        context.getContext().setAttribute(WebSocketContainer.class.getName(), container);
+        context.addManaged(container);
```

Comment:
why is the context directly managing the container?  Surely it is managed by the WSUHandler, which itself will be managed by the context?

It is kind of strange that this method does not insert the WSUHandler into the context?

Reply:
Now the container is managed by the WSUH.

Application code may want to insert another `Handler` between the `ContextHandler` and the WSUH, so it is not done here.